### ISSUE TITLE
Configurable ramp-up for concurrency

### ISSUE THREAD

**Issue by mhjort** [mhjort/clj-gatling#20]
At the moment clj-gatling starts the simulation with a given number of concurrent users and runs simulation constantly with the same number of users. In load testing it's often better to ramp-up users within a given time period. For example Gatling provides a huge set of [configurable algorithms](http://gatling.io/docs/2.1.1/general/simulation_setup.html) for specifying the ramp-up. 

clj-gatling philosophy is not to have a dsl but instead give the user a control for specifying her own algorithm. My proposal for this is following. Instead of configuring number of concurrent users as an integer the concurrency will be a function that gets a progress as a parameter and returns the number of users. Example:

Constant load with 100 users

``` clojure
:concurrency 
  (fn [_]
    (constantly 100))
```

Ramp up (first 10, then 50 and finally 100 users)

``` clojure
:concurrency
  (fn [progress-percentage]
    (cond 
      (< progress-percentage 5) 10
      (< progress-percentage 10) 50
      :else 100))
```

Does this make sense?

**Comment by yashaka** [comment]
Yep, this makes a lot of sense. 

And this example:

```
:concurrency 
  (fn [_]
    (constantly 100))
```

seems like can be simplified to the following:

```
:concurrency (constantly 100)
```

This also seems to give an ability to generate "load with a consecutive peaks" like:
```
     --        --        --        
    |  |      |  |      |  |      
____|  |______|  |______|  |______
```

which is rather valuable in context of "spike testing".

But I am still not sure, should we deprecate the "old" way...

*Option1*

Maybe the logic should be smart enough to check whether :concurrency is fn or not. And in case if not - convert it to `(constantly concurrency)`

In addition, if we leave the ability to specify concurrency as non-fn value, then maybe it makes sense to add `:ramp-up` option too... The logic can be the following: use `:ramp-up` in case `:concurrency` is not fn. 

We already have similar logic: ignore `:requests` in case `:duration` is set. So why not implement something similar for `:ramp-up`?

Actually the following are two the most popular cases in performance testing:

Load testing:
```
:concurrency 200
:ramp-up (clj-time.core/seconds 10)
:duration (clj-time.core/seconds 600)
``` 

Stress testing
```
:concurrency 1000
:ramp-up (clj-time.core/seconds 600)
:duration (clj-time.core/seconds 600)
``` 

So they will be used most often... 

*Option2* (:concurrency as fn only)
From other point of view such "smart logic" may be not KISS enough... 

Actually the same can be implemented like this:

Load testing:
```
:concurrency (-> 200 (with-ramp-up-percentage 1))
:duration (clj-time.core/seconds 600)
``` 

Stress testing
```
:concurrency (-> 1000 (with-ramp-up-percentage 100))
:duration (clj-time.core/seconds 600)
``` 

Where user defined `with-ramp-up-percentage` will increase concurrency (first parameter) gradually unless ramp-up in percentage (second parameter) is reached.

*Option3*
We also can leave everything as it is, where `:concurrency` is non-fn, add `:ramp-up` as time period, so examples from Option1 will work... 
And add `:concurrency-fn` with logic: if `:concurrency-fn` is set, ignore `:concurrency & :ramp-up`

Sorry for too much of text:)

**Comment by mhjort** [comment]
Sorry for getting back to this issue took me sooo long. And thanks @yashaka for your thoughts. They were really helpful. I think I will go with the function approach because it is most flexible. This is the philosophy I have had with this tool. Later on it is easy to implement ready-made functions for different ramp-up algorithms if needed. That said I also want to keep the backwards compatibility. So the feature would be following:

Constant load with 100 users

``` clojure
:concurrency  100
```

Ramp up (first 10, then 50 and finally 100 users)

``` clojure
:concurrency
  (fn [progress-percentage context] ;context will be passed to all the functions
    (cond 
      (< progress-percentage 5) 10
      (< progress-percentage 10) 50
      :else 100))
```

**Comment by mhjort** [comment]
There is a now version `[0.10.0-beta1]` if someone of you have time to test this. Here is an example on how to use this new functionality: https://github.com/mhjort/clj-gatling-example/blob/master/src/clj_gatling_example/core.clj

I have also created PR but it is still work-in-progress because I would like to improve the tests. https://github.com/mhjort/clj-gatling/pull/49

**Comment by mhjort** [comment]
Implemented in `0.10.0`
